src/uwtools/tests/test_rocoto.py

```diff
@@ -44,26 +46,69 @@ def test_add_jobname():
+    assert module.called_once_with(tasks)
+
+
+def test__rocoto_schema_yaml():
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        expected = path / "rocoto.jsonschema"
+    assert rocoto._rocoto_schema_yaml() == expected
+
+
+def test__rocoto_schema_xml():
+    with resources.as_file(resources.files("uwtools.resources")) as path:
+        expected = path / "schema_with_metatasks.rng"
+    assert rocoto._rocoto_schema_xml() == expected
+
+
+@pytest.mark.parametrize("vals", [("hello_workflow.yaml", True), ("fruit_config.yaml", False)])
+def test_realize_rocoto_xml(vals, tmp_path):
+    fn, validity = vals
     output = tmp_path / "rendered.xml"
-    rocoto.write_rocoto_xml(
-        input_yaml=input_yaml, input_template=str(input_template), rendered_output=str(output)
-    )
 
-    expected = support.fixture_path("hello_workflow.xml")
-    support.compare_files(expected, output)
+    with patch.object(rocoto, "validate_rocoto_xml", value=True):
+        with patch.object(rocoto.uwtools.config.validator, "_bad_paths", return_value=None):
+            with resources.as_file(resources.files("uwtools.tests.fixtures")) as path:
+                config_file = path / fn
+                result = rocoto.realize_rocoto_xml(config_file=config_file, rendered_output=output)
+    assert result is validity
+
+
+def test_realize_rocoto_invalid_xml():
+    config_file = support.fixture_path("hello_workflow.yaml")
+    xml = support.fixture_path("rocoto_invalid.xml")
+    with patch.object(rocoto, "_write_rocoto_xml", return_value=None):
+        with patch.object(rocoto.uwtools.config.validator, "_bad_paths", return_value=None):
+            with patch.object(tempfile, "NamedTemporaryFile") as context_manager:
+                context_manager.return_value.__enter__.return_value.name = xml
+                result = rocoto.realize_rocoto_xml(config_file=config_file, rendered_output=xml)
+    assert result is False
 
 
 @pytest.mark.parametrize("vals", [("hello_workflow.xml", True), ("rocoto_invalid.xml", False)])
 def test_rocoto_xml_is_valid(vals):
     fn, validity = vals
-    with resources.as_file(resources.files("uwtools.resources")) as resc:
-        with open(resc / "schema_with_metatasks.rng", "r", encoding="utf-8") as f:
-            schema = etree.RelaxNG(etree.parse(f))
-
     xml = support.fixture_path(fn)
-    tree = etree.parse(xml)
-    assert schema.validate(tree) is validity
+    result = rocoto.validate_rocoto_xml(input_xml=xml)
+
+    assert result is validity
+
+
+def test__write_rocoto_xml(tmp_path):
+    config_file = support.fixture_path("hello_workflow.yaml")
+    output = tmp_path / "rendered.xml"
+    config = YAMLConfig(config_file)
+
+    with patch.object(rocoto, "_add_jobname_to_tasks", return_value=config):
+        rocoto._write_rocoto_xml(config_file=config_file, rendered_output=output)
+
+    expected = support.fixture_path("hello_workflow.xml")
+    support.compare_files(expected, output)
```

Comment:
Do we maybe need to make this an `assert` statement to actually make it fail if the files do not compare? Can you force it to fail now by interfering with one of the files?

Reply:
Assert is a stronger test regardless, but it sure fails when I add that! It looks like our example log path in `hello_workflow.yaml` was not getting written to the XML, but also its not drilling down to the third level - we can get the first task and metatask names, but the nested task is not pulling the name - I'm working on this one

Reply:
Ha, just had to stop patching the jobname, thanks!